File: .NET Core/EncryptCredentials/EncryptCredentials/Services/AadService.cs
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
// ----------------------------------------------------------------------------

namespace EncryptCredentials.Services
{
	using EncryptCredentials.Models;
	using Microsoft.Extensions.Options;
	using Microsoft.Identity.Client;
	using System;
	using System.Linq;
	using System.Security;

	public class AadService
	{
		private readonly IOptions<AzureAd> azureAd;

		public AadService(IOptions<AzureAd> azureAd)  
		{
			this.azureAd = azureAd;
		}

		/// <summary>
		/// Generates and returns Access token
		/// </summary>
		/// <returns>AAD token</returns>
		public string GetAccessToken()
		{
			AuthenticationResult authenticationResult = null;
			if (azureAd.Value.AuthenticationMode.Equals("masteruser", StringComparison.InvariantCultureIgnoreCase))
			{
				// Create a public client to authorize the app with the AAD app
				IPublicClientApplication clientApp = PublicClientApplicationBuilder.Create(azureAd.Value.ClientId).WithAuthority(azureAd.Value.AuthorityUri).Build();
				var userAccounts = clientApp.GetAccountsAsync().Result;
				try
				{
					// Retrieve Access token from cache if available
					authenticationResult = clientApp.AcquireTokenSilent(azureAd.Value.Scope, userAccounts.FirstOrDefault()).ExecuteAsync().Result;
				}
				catch (MsalUiRequiredException)
				{
					SecureString password = new SecureString();
					foreach (var key in azureAd.Value.PbiPassword)
					{
						password.AppendChar(key);
					}
					authenticationResult = clientApp.AcquireTokenByUsernamePassword(azureAd.Value.Scope, azureAd.Value.PbiUsername, password).ExecuteAsync().Result;
				}
			}

			// Service Principal auth is the recommended by Microsoft to achieve App Owns Data Power BI embedding
			else
			{
				// For app only authentication, we need the specific tenant id in the authority url
				var tenantSpecificUrl = azureAd.Value.AuthorityUri.Replace("organizations", azureAd.Value.TenantId);

				// Create a confidential client to authorize the app with the AAD app
				IConfidentialClientApplication clientApp = ConfidentialClientApplicationBuilder
																				.Create(azureAd.Value.ClientId)
																				.WithClientSecret(azureAd.Value.ClientSecret)
																				.WithAuthority(tenantSpecificUrl)
																				.Build();
				// Make a client call if Access token is not available in cache
				authenticationResult = clientApp.AcquireTokenForClient(azureAd.Value.Scope).ExecuteAsync().Result;
			}

			return authenticationResult.AccessToken;
		}
	}
}

File: Python/Encrypt credentials/Encryption sample/services/aadservice.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

import msal
from flask import current_app as app
from utils import Utils


class AadService:

    def get_access_token():
        ''' Generates and returns Access token

        Returns:
            string: Access token
        '''

        config_result = Utils.validate_config(app)
        if config_result:
            raise Exception(config_result)

        authenticate_mode = app.config['AUTHENTICATION_MODE']
        tenant_id = app.config['TENANT_ID']
        client_id = app.config['CLIENT_ID']
        username = app.config['POWER_BI_USER']
        password = app.config['POWER_BI_PASS']
        client_secret = app.config['CLIENT_SECRET']
        scope = app.config['SCOPE_BASE']
        authority = app.config['AUTHORITY_URL']
        response = None

        try:
            if authenticate_mode.lower() == 'masteruser':

                # Create a public client to authorize the app with the AAD app
                clientapp = msal.PublicClientApplication(client_id, authority=authority)

                # Make a client call if Access token is not available in cache
                response = clientapp.acquire_token_by_username_password(username, password, scopes=scope)

            # Service Principal auth is recommended by Microsoft to achieve App Owns Data Power BI embedding
            else:
                authority = authority.replace('organizations', tenant_id)
                clientapp = msal.ConfidentialClientApplication(client_id, client_credential=client_secret, authority=authority)

                # Make a client call if Access token is not available in cache
                response = clientapp.acquire_token_for_client(scopes=scope)

            return response['access_token']

        except KeyError:
            raise Exception(response['error_description'])
        except Exception as ex:
            raise Exception('Error retrieving Access token\n' + str(ex))
